### PR TITLE
IAPWS-IF97

### PR DESCRIPTION
**Pull request by cpgr**
- Implementation of IAPWS Industrial Formulation 97 (for properties as a function of p and T) for all five regions in the phase space
- Includes backwards equation for volume as a function of p and T for region 3 to avoid iteration
- Provides density, internal energy, entropy, enthalpy, cp, cv, speed of sound, viscosity and thermal conductivity
- Tests for all IAPWS IF97 verification examples

Note:
- To test the properties, a specific test material was created that is only really useful for the automatic test suite
- Derivatives of several properties wrt p and T are available, but I'm not sure how to include them in the tests
- There is a bit of code duplication - I'm open to suggestions on how to simplify this
- Backwards equations for T(p, h) or T(p, s) have not been implemented

Refs #6972

### CODE REVIEW

**Comment by permcody** [comment]
Hmm, This has got to be at least the third version of steam tables. We really need to consolidate these. Ping @andrsd

File: modules/fluid_properties/include/userobjects/Water97FluidProperties.h [hunk 1]
```diff
@@ -0,0 +1,1123 @@
...
+ *
+ * Water viscosity as a function of density (kg/m^3) and temperature (K) from:
+ * Release on the IAPWS Formulation 2008 for the Viscosity of Ordinary Water Substance.
+ *
+ * Thermal conductivity as a function of density (kg/m^3) and temperature (K) from:
+ * Revised Release on the IAPS Formulation 1985 for the Thermal Conductivity of
+ * Ordinary Water Substance
+ */
+class Water97FluidProperties : public SinglePhaseFluidPropertiesPT
+
```
Comment (andrsd) [inline modules/fluid_properties/include/userobjects/Water97FluidProperties.h]: Remove this empty line

File: modules/fluid_properties/include/userobjects/Water97FluidProperties.h [hunk 2]
```diff
@@ -0,0 +1,1123 @@
...
+{
+public:
+  Water97FluidProperties(const InputParameters & parameters);
+  virtual ~Water97FluidProperties();
+
+ /**
+  * Water molar mass
+  * @return molar mass (kg/mol)
+  */
+ virtual Real molarMass() const;
```
Comment (andrsd) [inline modules/fluid_properties/include/userobjects/Water97FluidProperties.h]: Indent by 2 spaces

File: modules/fluid_properties/src/userobjects/Water97FluidProperties.C [hunk 3]
```diff
@@ -0,0 +1,1411 @@
...
+       tau = _T_star[4] / temperature;
+       Real dgdp = dgamma5_dpi(pi, tau);
+       Real d2gdpt = d2gamma5_dpitau(pi, tau);
+       dinternal_energy_dp = _Rw * temperature * (tau * d2gdpt - dgdp - pi *
+         d2gamma5_dpi2(pi, tau)) / _p_star[4];
+       dinternal_energy_dT = _Rw * (tau * tau * d2gamma5_dtau2(pi, tau) - pi *
+         dgdp - pi * tau * d2gdpt);
+       break;
+     }
+    }
```
Comment (andrsd) [inline modules/fluid_properties/src/userobjects/Water97FluidProperties.C]: The switch-statement should be indented by 2 spaces (you have 3)

File: modules/fluid_properties/src/userobjects/Water97FluidProperties.C [hunk 4]
```diff
@@ -0,0 +1,1411 @@
...
+     case 3:
+       density = densityRegion3(pressure, temperature);
+       break;
+
+     case 5:
+       pi = pressure / _p_star[4];
+       tau = _T_star[4] / temperature;
+       density = pressure / (pi * _Rw * temperature * dgamma5_dpi(pi, tau));
+       break;
+    }
```
Comment (andrsd) [inline modules/fluid_properties/src/userobjects/Water97FluidProperties.C]: The closing `{` needs to match the indentation of the switch, i.e. 2 spaces

File: modules/fluid_properties/src/userobjects/Water97FluidProperties.C [hunk 5]
```diff
@@ -0,0 +1,1411 @@
...
+       break;
+     }
+
+     case 5:
+       pi = pressure / _p_star[4];
+       tau = _T_star[4] / temperature;
+       internal_energy = _Rw * temperature * (tau * dgamma5_dtau(pi, tau)
+         - pi * dgamma5_dpi(pi, tau));
+       break;
+    }
```
Comment (andrsd) [inline modules/fluid_properties/src/userobjects/Water97FluidProperties.C]: The closing `{` needs to match the indentation of the switch, i.e. 2 spaces

File: modules/fluid_properties/src/userobjects/Water97FluidProperties.C [hunk 6]
```diff
@@ -0,0 +1,1411 @@
...
+   // Determine which region the point is in
+   unsigned int region = inRegion(pressure, temperature);
+   switch (region)
+   {
+     case 1:
+       pi = pressure / _p_star[0];
+       tau = _T_star[0] / temperature;
+       speed2 = _Rw * temperature * std::pow(dgamma1_dpi(pi, tau), 2.0) /
+       (std::pow(dgamma1_dpi(pi, tau) - tau * d2gamma1_dpitau(pi, tau), 2.0) /
+       (tau * tau * d2gamma1_dtau2(pi, tau)) - d2gamma1_dpi2(pi, tau));
```
Comment (andrsd) [inline modules/fluid_properties/src/userobjects/Water97FluidProperties.C]: Indent these 2 lines by 2 more spaces, so it is clear they are the continuation of the long expression from 3 lines above.

File: modules/fluid_properties/src/userobjects/Water97FluidProperties.C [hunk 7]
```diff
@@ -0,0 +1,1411 @@
...
+     }
+
+     case 5:
+       pi = pressure / _p_star[4];
+       tau = _T_star[4] / temperature;
+       speed2 = _Rw * temperature * std::pow(pi * dgamma5_dpi(pi, tau), 2.0) /
+         ((- pi * pi * d2gamma5_dpi2(pi, tau)) + std::pow(pi * dgamma5_dpi(pi, tau) -
+         tau * pi * d2gamma5_dpitau(pi, tau), 2.0)/(tau * tau * d2gamma5_dtau2(pi, tau)));
+       break;
+    }
```
Comment (andrsd) [inline modules/fluid_properties/src/userobjects/Water97FluidProperties.C]: The closing `{` needs to match the indentation of the switch, i.e. 2 spaces

File: modules/fluid_properties/src/userobjects/Water97FluidProperties.C [hunk 8]
```diff
@@ -0,0 +1,1411 @@
...
+         (2.0 * delta * dphi3_ddelta(delta, tau) + delta * delta * d2phi3_ddelta2(delta, tau)));
+       break;
+     }
+
+     case 5:
+       pi = pressure / _p_star[4];
+       tau = _T_star[4] / temperature;
+       specific_heat = - _Rw * tau * tau * d2gamma5_dtau2(pi, tau);
+       break;
+    }
```
Comment (andrsd) [inline modules/fluid_properties/src/userobjects/Water97FluidProperties.C]: The closing `{` needs to match the indentation of the switch, i.e. 2 spaces

File: modules/fluid_properties/src/userobjects/Water97FluidProperties.C [hunk 9]
```diff
@@ -0,0 +1,1411 @@
...
+     }
+
+     case 5:
+       pi = pressure / _p_star[4];
+       tau = _T_star[4] / temperature;
+       specific_heat = _Rw * (- tau * tau * d2gamma5_dtau2(pi, tau) +
+         std::pow(dgamma5_dpi(pi, tau) - tau * d2gamma5_dpitau(pi, tau), 2) /
+         d2gamma5_dpi2(pi, tau));
+       break;
+    }
```
Comment (andrsd) [inline modules/fluid_properties/src/userobjects/Water97FluidProperties.C]: The closing `{` needs to match the indentation of the switch, i.e. 2 spaces

File: modules/fluid_properties/src/userobjects/Water97FluidProperties.C [hunk 10]
```diff
@@ -0,0 +1,1411 @@
...
+       tau = _T_star[2] / temperature;
+       entropy = _Rw * (tau *  dphi3_dtau(delta, tau) - phi3(delta, tau));
+       break;
+
+     case 5:
+       pi = pressure / _p_star[4];
+       tau = _T_star[4] / temperature;
+       entropy = _Rw * (tau * dgamma5_dtau(pi, tau) - gamma5(pi, tau));
+       break;
+    }
```
Comment (andrsd) [inline modules/fluid_properties/src/userobjects/Water97FluidProperties.C]: The closing `{` needs to match the indentation of the switch, i.e. 2 spaces

File: modules/fluid_properties/src/userobjects/Water97FluidProperties.C [hunk 11]
```diff
@@ -0,0 +1,1411 @@
...
+         delta *  dphi3_ddelta(delta, tau));
+       break;
+     }
+
+     case 5:
+       pi = pressure / _p_star[4];
+       tau = _T_star[4] / temperature;
+       enthalpy = _Rw * _T_star[4] * dgamma5_dtau(pi, tau);
+       break;
+    }
```
Comment (andrsd) [inline modules/fluid_properties/src/userobjects/Water97FluidProperties.C]: This block should be indented by 2 spaces (3 here) and the closing `}` needs to match the indentation of the switch, i.e. 2 spaces

File: modules/fluid_properties/src/userobjects/Water97FluidProperties.C [hunk 12]
```diff
@@ -0,0 +1,1411 @@
...
+     }
+
+     case 5:
+       pi = pressure / _p_star[4];
+       tau = _T_star[4] / temperature;
+       enthalpy = _Rw * _T_star[4] * dgamma5_dtau(pi, tau);
+       denthalpy_dp = _Rw * _T_star[4] * d2gamma5_dpitau(pi, tau) / _p_star[4];
+       denthalpy_dT = - _Rw * tau * tau * d2gamma5_dtau2(pi, tau);
+       break;
+    }
```
Comment (andrsd) [inline modules/fluid_properties/src/userobjects/Water97FluidProperties.C]: Ident the block by 2 spaces and fix the indentation of the closing '}'.

File: modules/fluid_properties/src/userobjects/Water97FluidProperties.C [hunk 13]
```diff
@@ -0,0 +1,1411 @@
...
+
+  return sum;
+}
+
+Real
+Water97FluidProperties::d2gamma1_dpi2(Real pi, Real tau) const
+{
+  Real sum = 0.0;
+  for (unsigned int i = 0; i < _n1.size(); ++i)
+    sum += _n1[i] * _I1[i] *(_I1[i] - 1) * std::pow(7.1 - pi, _I1[i] - 2) * std::pow(tau - 1.222, _J1[i]);
```
Comment (andrsd) [inline modules/fluid_properties/src/userobjects/Water97FluidProperties.C]: missing a space after `*`.

File: modules/fluid_properties/src/userobjects/Water97FluidProperties.C [hunk 14]
```diff
@@ -0,0 +1,1411 @@
...
+    sum0 += _n02[i] * std::pow(tau, _J02[i]);
+
+  Real g0 = std::log(pi) + sum0;
+
+  // Residual part of the Gibbs free energy
+  Real gr = 0.0;
+  for (unsigned int i = 0; i < _n2.size(); ++i)
+  {
+    gr += _n2[i] * std::pow(pi, _I2[i]) * std::pow(tau - 0.5, _J2[i]);
+  }
```
Comment (andrsd) [inline modules/fluid_properties/src/userobjects/Water97FluidProperties.C]: drop the curlies `{`, `}`...

File: modules/fluid_properties/include/userobjects/Water97FluidProperties.h [hunk 15]
```diff
@@ -0,0 +1,1123 @@
...
+  /**
+   * Water viscosity and derivatives wrt density and temperature
+   *
+   * @param pressure fluid pressure (Pa)
+   * @param temperature fluid temperature (K)
+   * @param[out] mu viscosity (Pa.s)
+   * @param[out] dmu_drho derivative of viscosity wrt density
+   * @param[out] dmu_dT derivative of viscosity wrt temperature
+   */
+  virtual void mu_drhoT(Real denisty, Real temperature, Real & mu, Real & dmu_drho, Real & dmu_dT) const override;
```
Comment (andrsd) [inline modules/fluid_properties/include/userobjects/Water97FluidProperties.h]: s/denisty/density/

File: modules/fluid_properties/include/userobjects/Water97FluidProperties.h [hunk 16]
```diff
@@ -0,0 +1,1123 @@
...
+   * @param density fluid density (kg/m^3)
+   * @param temperature fluid temperature (K)
+   * @return viscosity (Pa.s)
+   */
+  virtual Real mu(Real density, Real temperature) const override;
+
+  /**
+   * Water viscosity and derivatives wrt density and temperature
+   *
+   * @param pressure fluid pressure (Pa)
```
Comment (andrsd) [inline modules/fluid_properties/include/userobjects/Water97FluidProperties.h]: `@param density fluid density (<your unit here>)`

File: modules/fluid_properties/src/userobjects/Water97FluidProperties.C [hunk 17]
```diff
@@ -0,0 +1,1411 @@
...
+      subregion = 23;
+  }
+  else
+    mooseError("Water97FluidProperties::subregion3. Shouldn't have got here!");
+
+  return subregion;
+}
+
+Real
+Water97FluidProperties::tempXY(Real pressure, std::string xy) const
```
Comment (andrsd) [inline modules/fluid_properties/src/userobjects/Water97FluidProperties.C]: You typically what to stay away from string comparisons in the code parts that are called by the most inner loops like this one. Unless you have really good reason for doing the 2-character string thing, this needs to be done differently. For example, you could build an enum like so:
```
enum ERowType {
  AB,
  CD,
  ...
}
```
Then you would pass in the enum and compared on that...

File: modules/fluid_properties/src/userobjects/Water97FluidProperties.C [hunk 18]
```diff
@@ -0,0 +1,1411 @@
...
+  else if (xy == "qu")
+    row = 7;
+  else if (xy == "rx")
+    row = 8;
+  else if (xy == "uv")
+    row = 9;
+  else if (xy == "wx")
+    row = 10;
+  else if (xy == "ef")
+    row = 0; // not used
```
Comment (andrsd) [inline modules/fluid_properties/src/userobjects/Water97FluidProperties.C]: Either the comment lies or you have something wrong, `row = 0` is used by `xy == "ab"`

Comment (andrsd) [inline modules/fluid_properties/src/userobjects/Water97FluidProperties.C]: I see now, "ef" does not require `row`... OK. Maybe not have that in the if-statement then?

**Comment by andrsd** [comment]
As for the syntax in input files. We are using the new syntax, i.e. top level block like so:
```
[ICs]
  [./p_ic]
    type = FunctionIC
    variable = pressure
    function = pic
  [../]
[]
```

**Comment by andrsd** [comment]
@permcody Yeah, but this one is purely C++ ;-) So we like it more. Moreover it is wrapped in our nice `FluidProperty` interface (we like that too ;-)).

**Comment by permcody** [comment]
Right - but we have the one in water_steam_eos and the one you are using internally. I'm just trying to figure out if we can get rid of one or more of these.

File: modules/fluid_properties/src/userobjects/Water97FluidProperties.C (hunk 17 above)
Comment (cpgr) [inline modules/fluid_properties/src/userobjects/Water97FluidProperties.C]: Yeah, my really good reason was that I was lazy and just cut and pasted what I had done a few years ago. I'll swap this over to an enum.

**Comment by cpgr** [comment]
I need water properties from (pressure, temperature), so IAPWS-IF97 is the logical choice. From what I understand, @andrsd uses IAPWS95 (with volume and energy as the variables), so we can't really get away from two IAPWS formulations. Also, I think that the IAPWS95 version isn't even available for others to use.

I didn't know about the water_steam_eos module when I wrote most of this implementation back in my first MOOSE app a few years ago. As far as I can see, water_steam_eos also uses IAPWS-IF97, but doesn't provide all of the properties that I want. I thought it would be easiest to just use what I previously had and put it into the FluidProperties module, rather than add to the existing water_steam_eos module.

I imagine that this could provide all the functionality of the water_steam_eos without much extra effort. This was really a cut and paste job to get a working version in the FluidProperties module, where we will eventually move the fluids that we have in PorousFlow module to.

**Comment by permcody** [comment]
I'm in full support of this effort. I just want to figure out if there's a
place to cooperate. I actually want to delete the water_steam_eos module so
I may work on making that happen sooner than later. I'll talk to David and
see if he can use this version or not. It would be cool to get to one of we
can or at least organize them together if we can.

**Comment by permcody** [comment]
@andrsd - I'm assigning this PR to you. Thanks

**Comment by andrsd** [comment]
The `fluid_properties` module should eventually replace the `water_steam_eos`. It can handle different formulations that are suitable for different equation sets. The conserved equations require (u, v) formulation (specific energy and specific volume), the equations with primitive variables require typically p, T (pressure, temperature). From what I recall, sometimes people would want p, H (pressure, enthalpy). With the current design, the `fluid_properties` module should be able to do this.

One thing I'd like to see is that the names of the fluid properties classes should reflect which formulation that are. For example, there exists IAPWS95 formulation in terms of p and T, but the one we need in RELAP-7 and bighorn has to be in terms of u and v. But we can make this change later... Just a heads up...

File: modules/fluid_properties/src/userobjects/Water97FluidProperties.C [hunk 19]
```diff
@@ -0,0 +1,1421 @@
...
+                        {0.584561202520006e3, -0.102961025163669e1, 0.243293362700452, -0.294905044740799e-2},
+                        {0.528199646263062e3, 0.890579602135307e1, -0.222814134903755, 0.286791682263697e-2},
+                        {0.728052609145380e1, 0.973505869861952e2, 0.147370491183191e2, 0.329196213998375e3, 0.873371668682417e3}};
+
+  // Choose the constants based on the string xy
+  unsigned int row;
+
+  switch (xy)
+  {
+    case 0: // AB
```
Comment (andrsd) [inline modules/fluid_properties/src/userobjects/Water97FluidProperties.C]: You want to do: `case AB:`, etc. Do not need that comment and you will have nice legible code ;-)

File: modules/fluid_properties/src/userobjects/Water97FluidProperties.C [hunk 20]
```diff
@@ -0,0 +1,1421 @@
...
+        (2.0 * delta * dphi3_ddelta(delta, tau) + delta * delta * d2phi3_ddelta2(delta, tau)));
+      break;
+    }
+
+    case 5:
+      pi = pressure / _p_star[4];
+      tau = _T_star[4] / temperature;
+      specific_heat = - _Rw * tau * tau * d2gamma5_dtau2(pi, tau);
+      break;
+    }
```
Comment (andrsd) [inline modules/fluid_properties/src/userobjects/Water97FluidProperties.C]: Indent by 2 spaces

File: modules/fluid_properties/src/userobjects/Water97FluidProperties.C (hunk 19 above)
Comment (cpgr) [inline modules/fluid_properties/src/userobjects/Water97FluidProperties.C]: Yep, looks much better now

**Comment by cpgr** [comment]
Thanks for looking at this @andrsd

**Pull request merged**